lightning/src/chain/mod.rs

```diff
@@ -9,6 +9,124 @@
+pub trait Watch: Send + Sync {
+	/// Keys needed by monitors for creating and signing transactions.
+	type Keys: ChannelKeys;
+
+	/// Watches a channel identified by `funding_txo` using `monitor`.
+	///
+	/// Implementations are responsible for watching the chain for the funding transaction along
+	/// with any spends of outputs returned by [`get_outputs_to_watch`]. In practice, this means
+	/// calling [`block_connected`] and [`block_disconnected`] on the monitor.
+	///
+	/// [`get_outputs_to_watch`]: channelmonitor/struct.ChannelMonitor.html#method.get_outputs_to_watch
+	/// [`block_connected`]: channelmonitor/struct.ChannelMonitor.html#method.block_connected
+	/// [`block_disconnected`]: channelmonitor/struct.ChannelMonitor.html#method.block_disconnected
+	fn watch_channel(&self, funding_txo: OutPoint, monitor: ChannelMonitor<Self::Keys>) -> Result<(), ChannelMonitorUpdateErr>;
+
+	/// Updates a channel identified by `funding_txo` by applying `update` to its monitor.
+	///
+	/// Implementations must call [`update_monitor`] with the given update. See
+	/// [`ChannelMonitorUpdateErr`] for invariants around returning an error.
+	///
+	/// TODO(jkczyz): Determine where this should go from e73036c6845fd3cc16479a1b497db82a5ebb3897.
+	///
+	/// In case of distributed watchtowers deployment, even if an Err is return, the new version
+	/// must be written to disk, as state may have been stored but rejected due to a block forcing
+	/// a commitment broadcast. This storage is used to claim outputs of rejected state confirmed
+	/// onchain by another watchtower, lagging behind on block processing.
+	///
+	/// [`update_monitor`]: channelmonitor/struct.ChannelMonitor.html#method.update_monitor
+	/// [`ChannelMonitorUpdateErr`]: channelmonitor/enum.ChannelMonitorUpdateErr.html
+	fn update_channel(&self, funding_txo: OutPoint, update: ChannelMonitorUpdate) -> Result<(), ChannelMonitorUpdateErr>;
+
+	/// Returns any monitor events since the last call. Subsequent calls must only return new
+	/// events.
+	fn release_pending_monitor_events(&self) -> Vec<MonitorEvent>;
+}
+
+/// The `Filter` trait defines behavior for indicating chain activity of interest pertaining to
+/// channels.
+///
+/// This is useful in order to have a [`Watch`] implementation convey to a chain source which
+/// transactions to be notified of. Notifying may take the form of pre-filtering blocks or, in the
+/// case of [BIP 157]/[BIP 158], only fetching a block if the compact filter matches. After a
+/// transaction or an output has been registered, subsequent block retrievals from the chain source
+/// must not exclude any transactions matching the new criteria.
+///
+/// Note that use as part of a [`Watch`] implementation involves reentrancy. Therefore, these
+/// functions should not block on I/O. Implementations should instead queue the newly monitored data
+/// to be processed later. This may also require returning [`TemporaryFailure`] from `Watch`
```

Comment:
To make sure I understand, when you say:

> should they limit it to when _called from_ a `Watch` function

Is "it" referring to "returning `TemporaryFailure`"?

And if so, are you saying they should only return `TemporaryFailure` from a `Watch` function if:
1. the `Watch` function invocation calls a `Filter` function (e.g. `register_output`) and
2. the `Filter` is waiting on the remote server's response from previously queued outputs to monitor?

If so -- aside from clearing up the docs -- would it be best to have the `Filter` functions signal this back as the return value? That would allow us to implement the logic for returning `TemporaryFailure` in `ChainMonitor`.  Otherwise, we would put the onus on the user to implement their own `Watch` in lieu of using the provided `ChainMonitor`. And such an implementation I'd expect would introduce a strong coupling between the user's `Watch` and `Filter` implementations.

Reply:
Yes, your understanding is correct - we could let `Filter` return the error, though its somewhat awkward to correctly communicate that this now requires calling `ChannelManager::channel_monitor_updated` which the `update_id` which was contained in the `ChannelMonitorUpdate` object (but presumably wasn't passed into the `Filter`? I suppose we could pass it all the way through but it feels like a layer violation (though of course does seem more usable?).

Reply:
Ah, I see. Taking a step back, I see the following situation when using `ChainMonitor` as the `chain::Watch`:
- `ChannelManager` calls `ChainMonitor::watch_channel` upon `funding_created`/`funding_signed`
- `ChainMonitor::watch_channel` registers the funding output with the user's `chain::Filter`, which it may queue for processing

Then either:
1. the same situation happens for another channel or
2. the user calls `ChainMonitor::block_connected`, which may register more outputs with the user's `chain::Filter`

There are no other situations where the `chain::Filter` is called from `ChainMonitor`, AFAICT.

Questions:
- Does (1) require any different action other than again returning `TemporaryFailure` unconditionally?
- Would the user wait until the remote server finishes processing before (2), i.e., calling `ChainMonitor::block_connected`?

Assuming the answers are no and yes, respectively, would it be reasonable for `ChainMonitor` to track `ChannelMonitorUpdate::update_id` and provide it through an interface to the user? This would allow the user to pass it to `ChannelManager::channel_monitor_updated` when the remote server has finished processing. Then the user wouldn't need to implement their own `chain::Watch` in this situation.

Anyhow, although this is related, I think I'm getting a little sidetracked. :) Going back to the original discussion on the documentation, I've re-worded the last sentence as:

> Then, in order to block until the data has been processed, any `Watch` invocation that has called the `Filter` must return `TemporaryFailure`.

Does that convey the idea sufficiently?

Reply:
Right, I think the new wording is fine. To answer your question, I think the answer is either A or B, but both are not required - the user could always Ok() true and just make the server do some kind of rescan dance if there's a race, or the user could always return TemporaryFailure + channel_monitor_updated and then never worry about rescans, but they don't need to do both.